streams/src/main/java/org/apache/kafka/streams/state/internals/LogicalKeyValueSegments.java

```diff
@@ -16,15 +16,33 @@
  */
 package org.apache.kafka.streams.state.internals;
 
+import java.util.HashMap;
+import java.util.Map;
 import org.apache.kafka.streams.processor.ProcessorContext;
 import org.apache.kafka.streams.processor.internals.ProcessorContextUtils;
 import org.apache.kafka.streams.state.internals.metrics.RocksDBMetricsRecorder;
 
+/**
+ * A {@link Segments} implementation which uses a single underlying RocksDB instance.
+ * Regular segments with {@code segmentId >= 0} expire according to the specified
+ * retention period. "Reserved" segments with {@code segmentId < 0} do not expire
+ * and are completely separate from regular segments in that methods such as
+ * {@link #getSegmentForTimestamp(long)}, {@link #getOrCreateSegment(long, ProcessorContext)},
+ * {@link #getOrCreateSegmentIfLive(long, ProcessorContext, long)},
+ * {@link #segments(long, long, boolean)}, and {@link #allSegments(boolean)}
+ * only return regular segments and not reserved segments. The methods {@link #flush()}
+ * and {@link #close()} flush and close both regular and reserved segments, due to
+ * the fact that both types of segments share the same physical RocksDB instance.
+ * To create a reserved segment, use {@link #createReservedSegment(long, String)} instead.
+ */
 public class LogicalKeyValueSegments extends AbstractSegments<LogicalKeyValueSegment> {
 
     private final RocksDBMetricsRecorder metricsRecorder;
     private final RocksDBStore physicalStore;
 
+    // reserved segments do not expire, and are tracked here separately from regular segments
+    private final Map<Long, LogicalKeyValueSegment> reservedSegments = new HashMap<>();
```

Comment:
Only one reserved segment is needed for the versioned store implementation (used for the latest value store) but I've chosen to implement support for reserved segments more generally in this class. If we think this is unnecessarily complex, I can update this to be a single `Optional<LogicalKeyValueSegment>` (or equivalent) instead.